network/local.go

```diff
@@ -277,27 +277,47 @@ func (lc *LocalConn) Type() ConnType {
 // All operations are thread-safe.
 // The messages are marshalled and stored in the queue as a slice of bytes.
 type connQueue struct {
-	*sync.Cond
-	queue  [][]byte
-	closed bool
+	incomingQueue chan []byte
+	outgoingQueue chan []byte
+	closeCh       chan bool
+	wg            sync.WaitGroup
 }
 
+// LocalMaxBuffer is the number of packets that can be sent simultaneously to the
+// same address.
+const LocalMaxBuffer = 200
+
 func newConnQueue() *connQueue {
-	return &connQueue{
-		Cond: sync.NewCond(&sync.Mutex{}),
+	cq := &connQueue{
+		incomingQueue: make(chan []byte, LocalMaxBuffer),
+		outgoingQueue: make(chan []byte, LocalMaxBuffer),
+		closeCh:       make(chan bool),
+	}
+	cq.wg.Add(1)
+	go cq.start()
+	return cq
+}
+
+func (c *connQueue) start() {
+	defer c.wg.Done()
+	for {
+		select {
+		case buff := <-c.incomingQueue:
+			c.outgoingQueue <- buff
+		case <-c.closeCh:
+			// to signal that the conn is closed
+			close(c.outgoingQueue)
+			close(c.incomingQueue)
+			return
+		}
 	}
 }
 
 // push inserts a packet in the queue.
-// push won't work if the connQueue is already closed and silently return.
+// push should not be called if the connQueue is closed. Since it's the
+// LocalManager's reponsability, it is enforced.
 func (c *connQueue) push(buff []byte) {
-	c.L.Lock()
-	defer c.L.Unlock()
-	if c.closed {
-		return
-	}
-	c.queue = append(c.queue, buff)
-	c.Signal()
+	c.incomingQueue <- buff
```

Comment:
I agree with you. And to push the idea further, we could even remove the `connQueue` struct entirely now that it is only comprised of channels and put *all* the functionality in the `LocalConn` struct.
 However, do we want to do this now / here ? I mean this is only to fix bugs, not to re-design again and again. I'm happy to do it later in a future PR if we make an issue for it. Because removing `connQueue` will also lead to a whole new serie of tests, scrutinization, etc and this PR will be merged in one month...